dataset/src/main/scala/frameless/functions/Lit.scala

```diff
@@ -8,16 +8,17 @@ import org.apache.spark.sql.types.DataType
 private[frameless] case class Lit[T <: AnyVal](
     dataType: DataType,
     nullable: Boolean,
-    toCatalyst: CodegenContext => ExprCode,
-    show: () => String
+    show: () => String,
+    catalystExpr: Expression // must be a generated Expression from a literal TypedEncoder's toCatalyst function
 ) extends Expression with NonSQLExpression {
   override def toString: String = s"FramelessLit(${show()})"
 
-  def eval(input: InternalRow): Any = {
+  lazy val codegen = {
```

Comment:
Any reason for making it a lazy val and not a regular function?

Reply:
it could get evaluated several times over potentially before a wholestage.  Well perhaps it doesn't after the foldable change.  Either way very little overhead from the lazy and much quicker than calling a second time.